File: leetcode/912. Sort an Array/912.js
// https://helloacm.com/javascript-coding-exercise-the-quicksort-implementation-in-javascript/
// https://leetcode.com/problems/sort-an-array/

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortArray = function(nums) {
    if (nums.length === 0) return [];
    // pick a random value e.g. random.choice
    let x = nums[Math.floor(Math.random() * nums.length)];
    let lt = [];
    let eq = [];
    let gt = [];
    // separate into three groups
    nums.map(a => {
       if (a < x) lt.push(a);
       else if (a === x) eq.push(a);
       else gt.push(a);
    });
    // concatenate the part results
    let n = sortArray(lt);
    n.push(...eq);
    n.push(...sortArray(gt));
    return n;
};

File: leetcode/912. Sort an Array/912.cpp
// https://helloacm.com/the-recursive-quicksort-implementation-in-c/
// https://leetcode.com/problems/sort-an-array/

class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        if (nums.empty()) return {};
        vector<int> lt;
        vector<int> eq;
        vector<int> gt;
        int p = nums[rand() % nums.size()];
        for (const auto &n: nums) {
            if (n < p) lt.push_back(n);
            else if (n == p) eq.push_back(n);
            else gt.push_back(n);
        }
        vector<int> lt1 = sortArray(lt);
        vector<int> gt1 = sortArray(gt);
        lt1.insert(end(lt1), begin(eq), end(eq));
        lt1.insert(end(lt1), begin(gt1), end(gt1));
        return lt1;
    }
};
